goosepaper/upload.py

```diff
@@ -17,51 +52,66 @@ def validateFolder(folder: str):
 
     return True
 
-def upload(filepath=None, replace=False, folder=None):
 
-    if not filepath:
-        parser = argparse.ArgumentParser("Upload Goosepaper to reMarkable tablet")
-        parser.add_argument(
-            "file", default=None, help="The file to upload",
-        )
-        args = parser.parse_args()
-        filepath = args.file
+def getallitems(client):
 
-    filepath = Path(filepath)
+    # So somehow I corrupted or broke my cloud during testing and any object
+    # which exists is getting returned twice in the object list from
+    # get_meta_items. Deleting items and re-adding them hasn't fixed it even
+    # using the official RM client. So this avoids that problem. I haven't found
+    # a real fix yet for my cloud though.
+    
+    allitems = [ item for item in client.get_meta_items() if (item.Parent != "trash") ]
 
-    client = Client()
+    items=[]
+    for tempitem in allitems:
+        if not any(item.ID == tempitem.ID for item in items):
+            items.append(tempitem)
+
+    return items
 
-    try:
-        client.renew_token()
-    except AuthError:
-        print(
-            "Looks like this if the first time you've uploaded, need to register the device"
-        )
-        print("Get the code from here: https://my.remarkable.com/connect/remarkable")
-        code = input()
-        print("registering")
-        client.register_device(code)
-        if not client.renew_token():
-            print("registration failed D:")
-        else:
-            print("registration successful")
+
+def upload(filepath, replace=False, folder=None):
+
+    client = auth_client()
+    
+    if not client: 
+        print ("Honk Honk! Couldn't auth! Is your rmapy configured?")
+        return False
 
     if not validateFolder(folder):
         return False
 
+    if not sanitycheck(folder,client):
+        return False
+
+    filepath = Path(filepath)
+
+    # Added error handling to deal with possible race condition where the file is mangled
+    # or not written out before the upload actually occurs such as an AV false positive.
+    # 'pdf' is a simple throwaway file handle to make sure that we retain control of the
+    # file while it's being imported.
+    
+    try:
+        with open (filepath.resolve()) as pdf:
+            doc = ZipDocument(doc=str(filepath.resolve()))
+    except IOError as err:
+        print (f"Error locating or opening {filepath}")
+        return False
+    
     paperCandidates = []
     paperFolder = None
 
-    items = client.get_meta_items()
-    for item in items:
+    for item in getallitems(client):
+
         # is it the folder we are looking for?
         if (folder and item.Type == "CollectionType"              # is a folder
-                and item.VissibleName == folder                   # has the name we're looking for
+                and item.VissibleName.lower() == folder.lower()   # has the name we're looking for
```

Comment:
Yes it does mean that. What I was trying to do was make sure that if the user had 'Newsletter.epub' as a separate file but then created 'newsletter.epub' goosepaper it wouldn't accidentally overwrite the first one. I'm not totally sure this is the best way to handle it though. It might be better to do a case insensitivity sanity check for Type==Document where the file names match regardless of case and throw an error?

Reply:
Also more confusingly for windows users, they don't expect case sensitivity so for them a lowercase test and uppercase test filename are equivalent.

Reply:
Ah that's an interesting point. I think this makes it _less_ likely that we accidentally overwrite a user's file, right? As long as we're being _overly_ cautious (even at the expense of not being able to have two weirdly-similarly-named-files), I think this is ok. Thanks for explaining!!